### PR TITLE
Example of Scalariform on Heimdal

### PR DESCRIPTION
**Pull request by andrewconner**
This PR shows what the default options (with one added as per recommendation from the Scala Style Guide) would do to Heimdal. This is set up to automatically run on every compile.

Probably best not to merge until we all agree on it.

@2is10 @lawzlo @stkem @raymondng @eishay @ymatsuda @leogrim @yingjieMiao @martinraison @ahsu1230

### CODE REVIEW

File: kifi-backend/modules/heimdal/app/com/keepit/controllers/AnalyticsController.scala [hunk 1]
```diff
@@ -3,20 +3,19 @@ package com.keepit.controllers
...
+import com.keepit.common.db.{ ExternalId, Id }
 import com.keepit.common.time._
 import com.keepit.model._
-import com.keepit.heimdal.{SpecificEventSet, _}
+import com.keepit.heimdal.{ SpecificEventSet, _ }
 import com.keepit.model.User
 import org.joda.time.DateTime
 import play.api.libs.concurrent.Execution.Implicits.defaultContext
-import play.api.libs.json.{JsNumber, JsObject, Json}
+import play.api.libs.json.{ JsNumber, JsObject, Json }
```
Comment (andrewconner) [inline kifi-backend/modules/heimdal/app/com/keepit/controllers/AnalyticsController.scala:13]: The style prefers spaces between `{` and `}` for the body.

File: kifi-backend/modules/heimdal/app/com/keepit/controllers/AnalyticsController.scala [hunk 2]
```diff
@@ -80,14 +79,13 @@ class AnalyticsController @Inject() (
     require(repo != SystemEvent.typeCode, s"Metrics are not supported yet for system events")
     if (request.queryString.get("help").nonEmpty) Ok(createHelp)
     else {
-      assert(window>0)
+      assert(window > 0)
```
Comment (andrewconner) [inline kifi-backend/modules/heimdal/app/com/keepit/controllers/AnalyticsController.scala:82]: Tokens are separated with spaces.

File: kifi-backend/modules/heimdal/app/com/keepit/controllers/AnalyticsController.scala [hunk 3]
```diff
@@ -24,7 +23,7 @@ class AnalyticsController @Inject() (
   anonymousEventLoggingRepo: AnonymousEventLoggingRepo,
   nonUserEventLoggingRepo: NonUserEventLoggingRepo,
   metricManager: MetricManager)
-  extends HeimdalServiceController {
+    extends HeimdalServiceController {
```
Comment (andrewconner) [inline kifi-backend/modules/heimdal/app/com/keepit/controllers/AnalyticsController.scala:26]: Class definitions use double indent when they span multiple lines, to provide a distinction between members of the class and the body. https://github.com/mdr/scalariform#doubleindentclassdeclaration

File: kifi-backend/modules/heimdal/app/com/keepit/controllers/AnalyticsController.scala [hunk 4]
```diff
@@ -102,13 +100,13 @@ class AnalyticsController @Inject() (
...
-    val names : Seq[String] = if (name=="all") {
-      Await.result(metricManager.getAvailableMetrics.map{ metricDescriptors => metricDescriptors.map(_.name)}, 20 seconds)
+    val names: Seq[String] = if (name == "all") {
+      Await.result(metricManager.getAvailableMetrics.map { metricDescriptors => metricDescriptors.map(_.name) }, 20 seconds)
     } else {
       name.split(",")
     }
 
-    val infoOptions : Seq[(Option[MetricDescriptor], String)] = Await.result(
+    val infoOptions: Seq[(Option[MetricDescriptor], String)] = Await.result(
```
Comment (andrewconner) [inline kifi-backend/modules/heimdal/app/com/keepit/controllers/AnalyticsController.scala:109]: Typing takes the form `variable: Type`

File: kifi-backend/modules/heimdal/app/com/keepit/controllers/AnalyticsController.scala [hunk 5]
```diff
@@ -116,13 +114,13 @@ class AnalyticsController @Inject() (
...
-    infoOptions.foreach{
+    infoOptions.foreach {
       case (Some(desc), name) => {
         val data = Await.result(metricManager.getMetric(name), 20 second)
-        if (data.isEmpty){
+        if (data.isEmpty) {
           errors = errors + "No Data for Metric: " + name + "\n"
         } else {
-          json = json.deepMerge{
+          json = json.deepMerge {
```
Comment (andrewconner) [inline kifi-backend/modules/heimdal/app/com/keepit/controllers/AnalyticsController.scala:123]: Blocks are separated by a space.

File: kifi-backend/modules/heimdal/app/com/keepit/heimdal/Metrics.scala [hunk 6]
```diff
@@ -176,14 +171,12 @@ class GroupedEventCountMetricDefinition(eventsToConsider: EventSet, contextRestr
...
 
-
-
-class GroupedUserCountMetricDefinition(eventsToConsider: EventSet, contextRestriction: ContextRestriction, groupField: EventGrouping, breakDown : Boolean = false, keySort : Boolean = false) extends SimpleMetricDefinition {
+class GroupedUserCountMetricDefinition(eventsToConsider: EventSet, contextRestriction: ContextRestriction, groupField: EventGrouping, breakDown: Boolean = false, keySort: Boolean = false) extends SimpleMetricDefinition {
   def aggregationForTimeWindow(startTime: DateTime, timeWindowSize: Duration): (Seq[PipelineOperator], Stream[BSONDocument] => Seq[BSONDocument]) = {
     val timeWindowSelector = Match(BSONDocument(
       "time" -> BSONDocument(
         "$gte" -> BSONDateTime(startTime.getMillis),
-        "$lt"  -> BSONDateTime(startTime.getMillis + timeWindowSize.toMillis)
```
Comment (andrewconner) [inline kifi-backend/modules/heimdal/app/com/keepit/heimdal/Metrics.scala:186]: Extra spaces are removed between tokens.

**Comment by eishay** [comment]
LGTM

**Comment by stkem** [comment]
dito, maybe allow extra spacing in the options, but I don't feel too strongly.

**Comment by leogrim** [comment]
:+1:

**Comment by raymondng** [comment]
Nice! As long as it's automated I am for it. Like Stephen I don't feel strongly about the specifics of the styling.

One potential issue: I do recall at times IntelliJ's scala plugin has troubles dealing with spaces. In general I don't like to see red lines everywhere in the IDE -- so will be keeping an eye on that for sure.

File: kifi-backend/modules/heimdal/app/com/keepit/controllers/AnalyticsController.scala (hunk 5 above)
Comment (raymondng) [inline kifi-backend/modules/heimdal/app/com/keepit/controllers/AnalyticsController.scala:123]: If memory serves I think this one might cause issues with IntelliJ's scala plugin.

File: kifi-backend/modules/heimdal/app/com/keepit/heimdal/MixpanelClient.scala [hunk 7]
```diff
@@ -72,14 +72,12 @@ class MixpanelClient(projectToken: String, shoebox: ShoeboxServiceClient) {
...
-
   private def sendData(url: String, data: JsObject) = {
     val request = WS.url(url).withQueryString(("data", Base64.encodeBase64String(Json.stringify(data).getBytes)))
     new SafeFuture(
       request.get().map {
         case response if response.body == "0\n" => throw new Exception(s"Mixpanel endpoint $url refused data: $data")
         case response => response
-      }
-    , Some("Mixpanel Event Forwarding"))
+      }, Some("Mixpanel Event Forwarding"))
```
Comment (2is10) [inline kifi-backend/modules/heimdal/app/com/keepit/heimdal/MixpanelClient.scala:81]: impressive

File: kifi-backend/modules/heimdal/app/com/keepit/model/EventRepo.scala [hunk 8]
```diff
@@ -24,11 +24,13 @@ trait EventAugmentor[E <: HeimdalEvent] extends PartialFunction[E, Future[Seq[(S
...
-        log.error(s"An augmentor failed on event: ${event.eventType}", ex)
-        Seq.empty
-      }
+    val safeAugmentations = augmentors.collect {
+      case augmentor if augmentor.isDefinedAt(event) =>
+        augmentor(event) recover {
+          case ex =>
+            log.error(s"An augmentor failed on event: ${event.eventType}", ex)
+            Seq.empty
+        }
```
Comment (2is10) [inline kifi-backend/modules/heimdal/app/com/keepit/model/EventRepo.scala:33]: wrapping a case is nice

**Comment by 2is10** [comment]
:+1: 

next: something sensible for imports?

**Comment by eishay** [comment]
can you publish the plan on moving forward with it?

**Pull request closed without merging**